### PR TITLE
Change aligner to generate cigar matches when bases are compatible

### PR DESCRIPTION
**Pull request by tfenne**
And also a tiny bit more performance tuning.

### CODE REVIEW

File: src/main/scala/com/fulcrumgenomics/alignment/Aligner.scala [hunk 1]
```diff
@@ -389,9 +395,15 @@ class Aligner(val scoringFunction: (Byte,Byte) => Int,
         currLength   = 1
       }
 
-      if (done()) elems += CigarElem(currOperator, currLength)
+      nextD = matrices(curD).trace(curI,curJ)
```
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/alignment/Aligner.scala]: space between commas

File: src/main/scala/com/fulcrumgenomics/alignment/Alignment.scala [hunk 2]
```diff
@@ -284,8 +274,22 @@ case class Alignment(query: Array[Byte],
...
+          op match {
+            case CigarOperator.EQ => alignBuffer.append(matchChar)
+            case CigarOperator.X  => alignBuffer.append(mismatchChar)
+            case CigarOperator.M  => alignBuffer.append(if (q == t) matchChar else mismatchChar)
+            case other            => throw new IllegalStateException(s"Unsupported cigar operator: $other")
+          }
+
+          qOffset += 1
+          tOffset += 1
+        }
```
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/alignment/Alignment.scala]: why was this moved to the last `case` statement?  I would think it happens the most frequently, so it would be best to have it first?

File: src/main/scala/com/fulcrumgenomics/util/Sequences.scala [hunk 3]
```diff
@@ -37,6 +37,54 @@ object Sequences {
   /** Common contig/chrom names for non-autosomal sequences in mammals. */
   val CommonNonAutosomalContigNames = Seq("M", "chrM", "MT", "X", "chrX", "Y", "chrY")
 
+  /** Array for determining what bases are compatible. Indices are all DNA/RNA bases. Contents
+    * are Int masks with a bit set for each possible base.  U is treated as identical to T.
+     */
```
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/util/Sequences.scala]: indent

File: src/main/scala/com/fulcrumgenomics/util/Sequences.scala [hunk 4]
```diff
@@ -37,6 +37,54 @@ object Sequences {
...
+    *   - `compatible('A', 'R') == true`
+    *   - `compatible('M', 'S') == true`
+    *   - `compatible('A', 'C') == false`
+    *   - `compatible('M', 'K') == false`
+    *
+    * @param base1 the first base to be compared
+    * @param base2 the second base to be compared
+    * @return true if the bases share at least one concrete base in common, false otherwise
+    */
+  final def compatible(base1: Byte, base2: Byte): Boolean = (IupacMasks(base1) & IupacMasks(base2)) > 0
```
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/util/Sequences.scala]: should `compatible('?'.toByte, ?'.toByte)` be allowed (i.e. should it throw an exception?  You could check if IupacMasks(base1/2) returns 0.  Up to you.

File: src/main/scala/com/fulcrumgenomics/alignment/Alignment.scala (hunk 2 above)
Comment (tfenne) [reply src/main/scala/com/fulcrumgenomics/alignment/Alignment.scala]: I had to either move it to the end, or turn it into three cases, because I want it to match all three of `=/X/M`, but also to know which one it was.  I've converted the code to match just on the operator so it should be more efficient, to make up for the change.

File: src/main/scala/com/fulcrumgenomics/util/Sequences.scala (hunk 4 above)
Comment (tfenne) [reply src/main/scala/com/fulcrumgenomics/util/Sequences.scala]: I waffled on this.  I've changed it to `base1 == base2 || ...` which will allow exact matches for characters we don't expect, and probably also make it more efficient when comparing mostly-identical sequences.
